File: 3Sum_BruteForce.py
# Created by Aashish Adhikari at 7:42 AM 1/9/2021

'''
Time Complexity:
O(n^3)

Space Complexity:
In the worst case, there will be n/3 unique triplets. So we store all of them in the hs. Hence, O(n)
'''

class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """

        hs = set()

        for i in range(len(nums)):
            for j in range(i+1, len(nums)):
                for k in range(j+1, len(nums)):

                    if nums[i] + nums[j] + nums[k] == 0:

                        sortedd = sorted([nums[i], nums[j], nums[k]]) #    <-- use tuple to hash since list is not hashable
                        tuplee = (sortedd[0], sortedd[1], sortedd[2])

                        if  tuplee not in hs:

                            hs.add(tuplee)

        return [item for item in hs]

File: 3SumBruteForce.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        if(nums == null || nums.length == 0) {
            return new ArrayList<>();
        }
        HashSet<List<Integer>>s = new HashSet<>();
        int n = nums.length;
        for(int i = 0; i < n; i++) {
            for(int j = i + 1; j < n; j++) {
                for(int k = j + 1; k < n; k++) {
                    if(nums[i] + nums[j] + nums[k] == 0) {
                        List<Integer> l = Arrays.asList(nums[i], nums[j], nums[k]);
                        Collections.sort(l);
                        s.add(l);
                    }
                }
            }
        }
        return new ArrayList<List<Integer>>(s);
    }
}